docs/smart_bits.md

```diff
@@ -0,0 +1,71 @@
+# SmartBits
+
+## Overview
+"Smart" types in magma (`SmartBits`, `SmartBit`) emulate verilog bits types, allowing very compact syntax for describing datapaths. For example, we can simply write the following:
+
+```python
+x = m.smart.SmartBits[7]()
+y = m.smart.SmartBits[9]()
+z = m.smart.SmartBits[12]()
+
+expr = (~(x + y) + z) << x
+
+a @= expr
+```
+
+Note that using raw magma types (i.e. `Bits`, `UInt`, `SInt`), we would have to do several explicit extensions and explicit promotion to `UInt` (for example, because addition is not defined on plain `Bits`). The smart types allow for writing this simple syntax while still compiling to raw magma types (and therefore the generated code still adheres to the typing rules of the raw magma types).
+
+It should also be noted that due to the implementation details, expressions on smart types are not actually operable magma types. In fact, the only thing you can do with them is wire them to other smart types, as in `a @= expr`. At the time of wiring, bit-widths and signedness of operands are resolved as a function of the expression itself and the left hand side of the assignment (see below for details). So, a statement like `b @= expr`, means that the generated RTL could different based on the bit-width of `b`.
+
+## Types
+There are two smart type constructors:
+* `SmartBits[bit_width, signed=False]`: returns a SmartBits type with bit-width `bit_width` and signedness `sign`. `sign` defaults to `False`.
```

Comment:
`SmartBits[bit_width, signed=False]`

(x) Doubt

Reply:
Sorry for the confusion, clarified the interface.